Add mcxw23 power management support

zejiang0jason opened the pull request:
Enabled modes:
      idle: SLEEP
      suspend: DEEP-SLEEP
      standby: POWER-DOWN with CPU retention
    
    OS Time Base: OSTIMER with fro1m clock source

ZhaoxiangJin commented on soc/nxp/mcx/mcxw/mcxw2xx/Kconfig.defconfig:
```diff
@@ -4,6 +4,9 @@
 
 if SOC_SERIES_MCXW2XX
 
+config MCUX_OS_TIMER
```
If PM enabled, ostimer will be the system timer? The OS timer clock frequency is 1MHz, that means the system in a low performance at active mode.
Ideally, the best approach is to use a combination of systick and ostimer, with ostimer used in low-power mode and systick used in active mode. see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/Kconfig.cortex_m_systick#L46

ZhaoxiangJin replied:
When SysTick is configured as the system clock, `CONFIG_SYS_CLOCK_TICKS_PER_SEC `is 32000000; when OSTimer is configured as the system clock, it is 1000000.

zejiang0jason replied:
`CONFIG_SYS_CLOCK_TICKS_PER_SEC` is 100, `SYS_CLOCK_HW_CYCLES_PER_SEC` is the clock source frequency.

ZhaoxiangJin commented on drivers/timer/mcux_os_timer.c:
```diff
@@ -179,12 +179,13 @@ static uint32_t mcux_lpc_ostick_compensate_system_timer(void)
+
+#if defined(CONFIG_MCUX_OS_TIMER_PM_POWERED_OFF)
+	/* Reset the OS Timer to a known state */
+	RESET_PeripheralReset(kOSEVENT_TIMER_RST_SHIFT_RSTn);
```
Outside this PR scope: calling the reset API would be the most appropriate approach here.

ZhaoxiangJin commented:
It is better to enable [power_mgmt_soc](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/pm/power_mgmt_soc) test

yeaissa commented on soc/nxp/mcx/mcxw/mcxw2xx/power.c:
This won't be working for BLE. The time the system can go to sleep is derived from the OS idle time and the BLEController idle time.
Here the state selected by the default policy is only considering the OS idle time, exit, min-residency latencies. 
A new policy shall be implemented to consider the controller IDLE time, and select the appropriate state depending on that.

zejiang0jason replied:
thanks, updated, considering BLE when select the next state.

yeaissa reviewed (commented):
Lets discuss this together, I see some missing points that needs to be considered for BLE.

zejiang0jason committed:
boards: nxp: mcxw23: Add power management MCXW23 board

Add mcxw23_evk and frdm_mcxw23 platforms to the power management
test suite to validate power management functionality on MCXW23
hardware.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

ZhaoxiangJin commented on soc/nxp/mcx/mcxw/mcxw2xx/power.c:
```diff
@@ -6,17 +6,58 @@
+	uint64_t wakeup_sources;
+
+	/* Set PRIMASK */
+	__disable_irq();
```
For `ARMV7_M`/`ARMV8_M` mainline, the `arch_irq_lock()` implementation is  `__set_BASEPRI_MAX(_EXC_IRQ_DEFAULT_PRIO);` while `ARMV6_M`/`ARMV8_M `baseline is `__disable_irq();`. So for  `ARMV7_M`/`ARMV8_M` mainline, we need `__disable_irq();` to set PRIMASK before configuring BASEPRI to prevent interruption before wake-up, just like [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/cortex_m/cpu_idle.c#L81).

zejiang0jason replied:
it is better to keep Zephyr code completed, no matter how hal_nxp has done.

yeaissa commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
```diff
@@ -32,12 +32,55 @@
+				compatible = "zephyr,power-state";
+				power-state-name = "suspend-to-idle";
+				min-residency-us = <1300>;
+				exit-latency-us = <400>;
```
I question this value, on SDK code this is set to 749.
https://github.com/nxp-mcuxpresso/mcux-devices-mcx/blob/main/MCXW/MCXW236/drivers/power_manager_mcxw23/device/fsl_pm_device.c#L31

yeaissa replied:
Ca we check for other states

yeaissa commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
I see os_timer state is still set to <disabled>, please remove.
I noticed that this init was missing and the timer was not behaving as expected:
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/mcx/mcxw/mcxw2xx/soc.c#L100

zejiang0jason replied:
because of this setting: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/Kconfig.mcux_os#L8, once os_timer is enabled in device tree, it will be set to `y` in kconfig, and be used as OS tick. I think it is better that only use it as OS tick in when power management is used, so my proposed way is only enable os_timer using dts overlay in specific project.

zejiang0jason commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
```diff
@@ -32,12 +32,55 @@
+			};
+
+			/* Suspend mode maps to Deep Sleep mode. */
+			suspend: suspend {
```
The code is added because ostimer can't work in standby mode for some SOCs, but it work in MCXW23 deep sleep and power down mode, so that code is not necessary.

zejiang0jason committed:
soc: nxp: mcxw2xx: Add poweroff support

Add poweroff MCXW2xx SoCs, support btn_wk pin wakeup.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

zejiang0jason commented on soc/nxp/mcx/mcxw/mcxw2xx/Kconfig.defconfig:
enabled in [soc/nxp/mcx/mcxw/mcxw2xx/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/98986/files#diff-32fc40b35cec6400cb7a1a8edd4bd32c88838d8e78e8bdf3b0fc3ec6fe1e0244) when PM is used.

ZhaoxiangJin commented on drivers/interrupt_controller/intc_nxp_pint.c:
```diff
@@ -175,6 +175,28 @@ void nxp_pint_pin_unset_callback(uint8_t pin)
+ *
+ * @param pin: The pin to get the PINT slot index for
+ * @return The allocated slot index, if not allocated, return -1
+ */
```
same doxygen comments (compare with the .h) can remove.

zejiang0jason committed:
soc: nxp: mcxw2xx: Enable the power management

Enabled modes:
  idle: SLEEP
  suspend: DEEP-SLEEP
  standby: POWER-DOWN with CPU retention

OS Time Base: OSTIMER with 32K clock source

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

ZhaoxiangJin commented on soc/nxp/mcx/mcxw/mcxw2xx/Kconfig.defconfig:
```diff
@@ -4,6 +4,13 @@
+# MCXW23 uses SYSTICK timer as the kernel timer by default.
+# In case we need to switch to OS timer, then
+# define CONFIG_MCUX_OS_TIMER=y
+config MCUX_OS_TIMER
```
should be `configdefault MCUX_OS_TIMER`, for the other redefined Kconfig options in this file, corresponding updates should be made as well.

JimmyChen-NXP commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -948,6 +976,69 @@ static int dma_mcux_lpc_get_attribute(const struct device *dev, uint32_t type, u
+			virtual_channel++;
+		}
+	}
+}
```
To restore/backup the register context, is it a special case for the DMA peripheral or it is common case for all peripherals? If it apply for all other peripherals, it looks need either a common functionality or it is really necessary?
